File: Algorithms/414. Third Maximum Number/javascript/solution.set-and-delete.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var thirdMax = function (nums) {
    nums = new Set(nums);
    let maximum = Math.max(...nums);

    if (nums.size < 3) {
        return maximum;
    }

    nums.delete(maximum);
    let secondMaximum = Math.max(...nums);
    nums.delete(secondMaximum);
    return Math.max(...nums);
};

File: Algorithms/414. Third Maximum Number/golang/solution.set-and-delete.go
func thirdMax(nums []int) int {
	setNum := make(map[int]bool)
	var maximum int = -int(^uint(0)>>1) - 1
	for _, num := range nums {
		setNum[num] = true
		if num > maximum {
			maximum = num
		}
	}

	if len(setNum) < 3 {
		return maximum
	}

	delete(setNum, maximum)
	var secondMaximum int = max(setNum)
	delete(setNum, secondMaximum)
	return max(setNum)
}

func max(nums map[int]bool) int {
	var maximum int = -int(^uint(0)>>1) - 1
	for num := range nums {
		if num > maximum {
			maximum = num
		}
	}
	return maximum
}